Services/ReplayFileService.cs

```diff
@@ -0,0 +1,174 @@
+                throw new ExternalDependencyException("Failed to delete the replay file", e);
+            }
+        }
+
+        /// <summary>
+        /// Get a device model script.
+        /// </summary>
+        public async Task<DataFile> GetAsync(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                this.log.Error("Simulation script id cannot be empty!");
+                throw new InvalidInputException("Simulation script id cannot be empty! ");
+            }
+
+            IDataRecord item;
+            try
+            {
+                item = await this.replayFilesStorage.GetAsync(id);
+            }
+            catch (ResourceNotFoundException)
+            {
+                throw;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to load replay file from storage", () => new { id, e });
+                throw new ExternalDependencyException("Unable to load device model script from storage", e);
+            }
+
+            try
+            {
+                var deviceModelScript = JsonConvert.DeserializeObject<DataFile>(item.GetData());
+                deviceModelScript.ETag = item.GetETag();
+                return deviceModelScript;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to parse device model script loaded from storage", () => new { id, e });
+                throw new ExternalDependencyException("Unable to parse device model script loaded from storage", e);
+            }
+        }
+
+        /// <summary>
+        /// Create a device model script.
+        /// </summary>
+        public async Task<DataFile> InsertAsync(DataFile replayFile)
+        {
+            replayFile.Created = DateTimeOffset.UtcNow;
+            replayFile.Modified = replayFile.Created;
+
+            if (string.IsNullOrEmpty(replayFile.Id))
+            {
+                replayFile.Id = Guid.NewGuid().ToString();
+            }
+
+            this.log.Debug("Creating a new replay file.", () => new { replayFile });
+
+            try
+            {
+                IDataRecord record = this.replayFilesStorage.BuildRecord(replayFile.Id,
+                    JsonConvert.SerializeObject(replayFile));
+
+                var result = await this.replayFilesStorage.CreateAsync(record);
+
+                replayFile.ETag = result.GetETag();
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Failed to insert new replay file into storage",
+                    () => new { replayFile, e });
+                throw new ExternalDependencyException(
+                    "Failed to insert new replay file into storage", e);
+            }
+
+            return replayFile;
+        }
+
+        /// <summary>
+        /// Validate replay file
+        /// </summary>
+        public string Validate(Stream stream)
```

Comment:
It sounds like this method reads the file and does validation. Should this method's name reflect that? It's currently named 'Validate', but that could be considered misleading as it also gets the content.

Reply:
Renamed

Reply:
The new name doesn't really capture that this method returns the file contents. Perhaps call this method 'ValidateAndReadContents'?